### ISSUE TITLE
CLI clean up

### ISSUE THREAD

**Issue by pvdrz** [memorysafety/sudo-rs#129]
Since we decided we don't need to be a perfect, in-place replacement for sudo, we should consider removing some of the CLI options that we added. I'm opening this issue to keep track of this discussion.

- [ ] `-A, --askpass`
- [ ] `-B, --bell`
- [ ] `-b, --background`
- [ ] `-C num, --close-from=num`
- [ ] `-D directory, --chdir=directory`
- [ ] `-E, --preserve-env`
- [ ] `--preserve-env=list`
- [ ] `-e, --edit`
- [ ] `-g group, --group=group`
- [ ] `-H, --set-home`
- [ ] `-h, --help`
- [ ] `-h host, --host=host`
- [ ] `-i, --login`
- [ ] `-K, --remove-timestamp`
- [ ] `-k, --reset-timestamp`
- [ ] `-l, --list`
- [ ] `-N, --no-update`
- [ ] `-n, --non-interactive`
- [ ] `-P, --preserve-groups`
- [ ] `-p prompt, --prompt=prompt`
- [ ] `-R directory, --chroot=directory`
- [ ] `-S, --stdin`
- [ ] `-s, --shell`
- [ ] `-U user, --other-user=user`
- [ ] `-T timeout, --command-timeout=timeout`
- [ ] `-u user, --user=user`
- [ ] `-V, --version`
- [ ] `-v, --validate`
- [ ] `--`

**Comment by allisonkarlitskaya** [comment]
hi,

Cockpit uses sudo to gain root privileges for users performing actions via the web interface.  We definitely need askpass support: we forward the prompt to the user, asking for the password, and passing it back.  We can't do that with a pseudoterminal forwarded to the user because we also support caching the login password and using that.

See https://github.com/cockpit-project/cockpit/issues/22305 for the original report on our side.

Thanks!

**Comment by squell** [comment]
Hi! Thanks for requesting this. Could you open this as a separate issue? That makes it easier to track.

**Comment by gramakri** [comment]
Ubuntu 25 switched out sudo to sudo-rs automatically for some reason and many of our tests started failing. Investigating further, this was because -E and -C are gone from sudo-rs.

The workaround to go back is `sudo update-alternatives --set sudo /usr/bin/sudo.ws`

We have a node.js program that runs as normal user and spins out worker processes. These worker processes have to run as root and we use sudo for this. In node, the IPC mechanism uses fd 4 (see https://nodejs.org/api/child_process.html). We pass -C 4 to sudo to prevent sudo from closing the IPC. This way the child worker process can send progress messages to the parent. What's the alternate mechanism, if you remove this flag?

We use -E to preserve envs across parent node.js and programs and the spun out program. But it seems `--preserve-env=list` is implemented, is it? In that case, we should be able to adjust the code to whitelist the exact env vars that the child processes need.

**Comment by bjorn3** [comment]
> We use -E to preserve envs across parent node.js and programs and the spun out program. But it seems --preserve-env=list is implemented, is it? In that case, we should be able to adjust the code to whitelist the exact env vars that the child processes need.

Indeed. We fully support whitelisting individual env vars. Only a blanket allow of all env vars is not supported.

> We have a node.js program that runs as normal user and spins out worker processes. These worker processes have to run as root and we use sudo for this. In node, the IPC mechanism uses fd 4 (see https://nodejs.org/api/child_process.html). We pass -C 4 to sudo to prevent sudo from closing the IPC. This way the child worker process can send progress messages to the parent. What's the alternate mechanism, if you remove this flag?

We don't blindly copy every feature of the original sudo for security reasons. Whenever we add a new feature we evaluate them based on the following criteria:

https://github.com/trifectatechfoundation/sudo-rs/blob/f2b9b996bf43c62d470a565e90b7e7f33769f158/CONTRIBUTING.md?plain=1#L18-L22

That said, I will discuss if we want to support `--close-from` with the rest of the sudo-rs team taking your use case into account.
